experimental/trusted_shuffler/trusted_shuffler/src/tests.rs

```diff
@@ -0,0 +1,135 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{RequestHandler, TrustedShuffler};
+use async_trait::async_trait;
+use futures::future::join_all;
+use std::{sync::Arc, time::Duration};
+use tokio::time::timeout;
+
+const TEST_TIMEOUT_MILLISECONDS: u64 = 100;
+
+struct TestRequestHandler;
+
+#[async_trait]
+impl RequestHandler for TestRequestHandler {
+    // A test function that processes requests.
+    // In the real use-case, this function should send requests to the backend and
+    // return responses.
+    async fn handle(&self, request: Vec<u8>) -> Vec<u8> {
+        generate_response(request)
+    }
+}
+
+// Non-async version of the `handle_request` function used to create expected responses.
+fn generate_response(request: Vec<u8>) -> Vec<u8> {
+    let parsed_request = String::from_utf8(request).expect("Couldn't parse request");
+    format!("Response for: {}", parsed_request)
+        .to_string()
+        .as_bytes()
+        .to_vec()
+}
+
+// Generates a request and a corresponding response from a string.
+fn generate_request_response(data: &str) -> (Vec<u8>, Vec<u8>) {
+    let request = format!("Request: {}", data).to_string().as_bytes().to_vec();
+    let response = generate_response(request.clone());
+    (request, response)
+}
+
+#[tokio::test]
+async fn trusted_shuffler_k2_test() {
+    let anonymity_value = 2;
+    let trusted_shuffler = Arc::new(TrustedShuffler::new(
+        anonymity_value,
+        Arc::new(TestRequestHandler {}),
+    ));
+
+    let (request, expected_response) = generate_request_response("Test");
+    let (background_request, expected_background_response) =
+        generate_request_response("Background test");
+
+    let trusted_shuffler_clone = trusted_shuffler.clone();
+    let background_result =
+        tokio::spawn(async move { trusted_shuffler_clone.invoke(background_request).await });
```

Comment:
Is it really necessary to put the trusted shuffler in an Arc, then use it in an async closure, and move the Arc into that? I would have thought that it's enough to just get the future out of it and spawn the future itself only.